tls/s2n_cipher_suites.c

```diff
@@ -1170,7 +1170,24 @@ static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, u
 }
 
 bool s2n_cipher_suite_uses_chacha20_alg(struct s2n_cipher_suite *cipher_suite) {
-    return cipher_suite && cipher_suite->record_alg && cipher_suite->record_alg->cipher == &s2n_chacha20_poly1305;
+    if (!cipher_suite) {
+        return false;
+    }
+
+    /* If chacha20-poly1305 is available, then we can directly check the record_alg */
+    if (s2n_chacha20_poly1305.is_available()) {
+        return cipher_suite->record_alg && cipher_suite->record_alg->cipher == &s2n_chacha20_poly1305;
+    } 
+
+    /* Else chacha20 poly1305 is not available, then the record_alg is never set in s2n_cipher_suites_init(). */
+    for (size_t i = 0; i < cipher_suite->num_record_algs; i++) {
+        const struct s2n_record_algorithm* rec_alg = cipher_suite->all_record_algs[i];
+        if (rec_alg && rec_alg->cipher == &s2n_chacha20_poly1305) {
+            return true;
+        }
+    }
```

Comment:
This seems like an unnecessary complication. You only need this part of the check for s2n_security_policies_init, not for the standard / runtime case. We don't even care if it's true at runtime-- we can't choose a cipher suite without a cipher.

Why not just skip your check in s2n_security_policies_init that security policies with chacha20 boosting enabled have a chacha20 cipher suite if chacha20 isn't supported? I remember you had a unit test to prove the logic behaved reasonably even with an "invalid" policy. 

If you absolutely think that check is necessary for all builds, then you need a comment explaining why this part of the check is necessary. Or you could move it into s2n_security_policies_init to make the "why" clearer.

Reply:
Yeah you're right. It feels weird to me to allow the init to pass, even though we know the chacha20 configuration doesn't make sense. I'm leaning towards moving the logic with `all_records_algs` to the sec policy init file.